### ISSUE TITLE
Create callback_list in drcontainers

### ISSUE THREAD

**Issue by byron-hawkins** [DynamoRIO/dynamorio#1762]
The new Dr. Fuzz feature of Dr. Memory (xref drmem [#1734](https://github.com/DynamoRIO/drmemory/issues/1734)) needs a generic callback list. Since drmgr already has one, we can just move it to drcontainers and make it available to all clients. 

A couple of changes should be made to the implementation to make it more friendly for general use:

1. In the add function, when the `before` priority is not `NULL`, the list should return an error if that priority is not found in the list
2. The resize function should remove invalid entries (i.e., not copy them to the new array) to minimize fragmentation for frequently modified callback lists

To avoid complications in the remove function, the list will need to allow direct index-based iteration, since existing usage by drmgr does not have a single key per entry (see `drmgr_bb_cb_remove()` in `drmgr.c`).

**Comment by byron-hawkins** [comment]
There's another significant issue with the current callback list in drmgr. The following scenario rejects the third callback registration, but it is satisfiable. There will be no way for the client to add that third callback in the correct place:

1. Add priority A (NULL before/after)
2. Add priority B (NULL before/after)
3. Add priority C, after B, before A

My new callback list will sort the nodes into {B,C,A}, but the current version in drmgr will return error. This seems very treacherous to me. In my test, the majority of valid priority orderings involving 3 and 4 registrations were rejected arbitrarily for this reason. If we really can't tolerate the complexity of a 90-line topological sort, we should at least warn the user that using multiple named priorities usually does not work.

**Comment by derekbruening** [comment]
> The resize function should remove invalid entries (i.e., not copy them to the new array) to minimize  fragmentation for frequently modified callback lists

I don't know that we need to spend much time worrying about a lot of dynamic callback reg/unreg: we have never written nor seen a tool that does that.

**Comment by derekbruening** [comment]
> There will be no way for the client to add that third callback in the correct place:
> 
> Add priority A (NULL before/after)
> Add priority B (NULL before/after)
> Add priority C, after B, before A
> My new callback list will sort the nodes into {B,C,A}, but the current version in drmgr will return error. 

First, you're not including the numeric priority, which is the *primary* sorter.  If you mean they all use 0 or something, then this *should* return error -- this is user error.  The priorities are not some shapeless soup you throw everything into that then gets sorted by these before/after fields: they have a defined order by numeric priority and registration order.  You can't insert something in between two other priorities when those two others are not specified to be in the order you want.

**Comment by byron-hawkins** [comment]
The numeric priority is currently the fallback option.

    /* Priority 3: numeric priority.  Tie goes to 1st to register. */    

Also the comment at the top of `priority_event_add()` says:

     /* priority can be NULL in which case default_priority is used. */

That default priority has numeric value 0. So the cases I mentioned are absolutely valid use of the API.

**Comment by byron-hawkins** [comment]
I recommend adding a new data structure `priority_list` to drcontainers that handles general priority lists, not just callbacks. It can be iterable without lock, and will require lock for add and remove. We can easily change drmgr to use the new priority list, and several components of drfuzz will use it too. The code is already written and thoroughly tested (all permutations before/after priorities for 4 registrations, excluding all cyclic cases).

**Comment by derekbruening** [comment]
> So the cases I mentioned are absolutely valid use of the API.

No, it is *invalid* to request to be inserted in between two callbacks that were originally added in the opposite order.  That is user error.

**Comment by byron-hawkins** [comment]
In that case I will simply remove support for multiple registrations from the drfuzz API.

**Comment by derekbruening** [comment]
Originally we only had numeric orderings, and we added the before and after as hints.

In order for your scenario to work, you have to *change the callback order* later on.  That means that originally A is called first and then B.  Once C registers and wants to be after B but before A, you now have to *reorder* A and B.  This is a different concept from how we've been thinking about this all along.  This could break any existing users who are depending on the static ordering for ties of registration order (which I thought was documented -- but I don't see it now).

I suppose we could switch to a dynamic ordering (documenting the change in the release notes) though to me it does not seem to be higher priority than many other features we'd like to have.

**Comment by zhaoqin** [comment]
I would suggest to remove the support for multiple registrations from the drfuzz API, we can add it later to drfuzz_mgr_* API or something, similar to DR's API and drmgr API.
So far, we have not seen any real use case of the multiple registrations, and we have other features that might be more important to work on.

**Comment by derekbruening** [comment]
It is true that the public docs are ambiguous and could be interpreted as having dynamic priorities.  As mentioned, that is not how at least I had been thinking about them.  I would be ok with changing to dynamic.

**Comment by derekbruening** [comment]
Xref https://codereview.appspot.com/257120043/diff/160001/common/utils_shared.c

For now we are clarifying in the drmgr docs and code that the numeric priority is the primary sorter and that the before and after are in some sense sanity checks on the numbers being right.
